code/game/objects/items/devices/uplinks.dm

```diff
@@ -239,6 +295,87 @@ GLOBAL_LIST_EMPTY(world_uplinks)
 			var/datum/uplink_item/UI = uplink_items[params["item"]]
 			return buy(UI, UI ? UI.reference : "")
 
+		if("add_to_cart")
+			var/datum/uplink_item/UI = uplink_items[params["item"]]
+			if(LAZYIN(shopping_cart, params["item"]))
+				to_chat(ui.user, "<span class='warning'>[UI.name] is already in your cart!</span>")
+				return
+			var/startamount = 1
+			if(UI.limited_stock == 0)
+				startamount = 0
+			LAZYSET(shopping_cart, params["item"], startamount)
+			generate_tgui_cart(TRUE)
+
+		if("remove_from_cart")
+			remove_from_cart(params["item"])
+
+		if("set_cart_item_quantity")
+			var/amount = text2num(params["quantity"])
+			LAZYSET(shopping_cart, params["item"], max(amount, 0))
+			generate_tgui_cart(TRUE)
+
+		if("purchase_cart")
+			if(!LAZYLEN(shopping_cart)) // sanity check
+				return
+			if(calculate_cart_tc() > uses)
+				to_chat(ui.user, "<span class='warning'>[src] buzzes, it doesn't contain enough telecrystals!</span>")
+				return
+			if(is_jammed)
+				to_chat(ui.user, "<span class='warning'>[src] seems to be jammed - it cannot be used here!</span>")
+				return
+
+			// Buying of the uplink stuff
+			var/list/bought_things = list()
+			for(var/reference in shopping_cart)
+				var/datum/uplink_item/item = uplink_items[reference]
+				var/purchase_amt = shopping_cart[reference]
+				if(purchase_amt <= 0)
+					continue
+				bought_things += mass_purchase(item, item ? item.reference : "", purchase_amt)
+
+			// Check how many of them are items
+			var/list/obj/item/items_for_crate = list()
+			for(var/thing in bought_things)
+				// because sometimes you can buy items like crates from surpluses and stuff
+				// the crates will already be on the ground, so we dont need to worry about them
+				if(isitem(thing))
+					items_for_crate += thing
```

Comment:
```suggestion
			for(var/obj/item/thing in bought_things)
				// because sometimes you can buy items like crates from surpluses and stuff
				// the crates will already be on the ground, so we dont need to worry about them
				items_for_crate += thing
```